Teleport Operator

marcoandredinis opened the pull request:
This PR adds a Kubernetes operator for Teleport: Teleport Operator.

You can check the `operator/README.md` to test the Operator yourself.

Note for reviewers:

There are a lot of file changes, but most of them are auto-generated.
Looking at each commit individually should ease the review process

Interesting files:
- `crdgen`: generates CRDs from the current API
- `sidecar`: creates a Teleport Client using a specific User and Role
- `controllers/resources`: reconcilers implementation

There's an extra pipeline in `.drone.yaml` which will be removed when the PR is approved.
It exists to test the steps we added to the current publish pipeline
It also provides a container image we can use to test out the Operator

RFD: https://github.com/gravitational/teleport-plugins/blob/master/rfd/0001-kubernetes-manager.md

marcoandredinis committed:
Add Teleport operator

This commit adds the Teleport operator. The operator reconciles
TeleportUsers and TeleportRoles Kubernetes resources with Users and
Roles Teleport resources.

strideynet commented on operator/crdgen/main.go:
```diff
@@ -0,0 +1,110 @@
+	"sigs.k8s.io/yaml"
+)
+
+func main() {
```
How challenging would it be to introduce a test that runs this against a super simple proto, and compares it against some generated files in `testdata/` ?

marcoandredinis replied:
Please take a look at https://github.com/gravitational/teleport/pull/13331/commits/ebe20548bbad221f6ebfcd76e1538484fff9b2f8

strideynet commented on operator/README.md:
```diff
@@ -0,0 +1,150 @@
+
+The first two steps above may change the object's state in K8S. If they do, we update and finish the cycle - we'll receive another reconciliation request with the new state.
+
+### Diagram
```
I'd probably label this diagram with "Pod" above the larger box, to make it clearer that the operator runs as a sidecar container in the teleport pod.

I'd maybe also clarify "k8s" as "kube-apiserver".

It's been a while since I've worked with Kubernetes controllers, but as far as I can remember, "k8s" doesn't really maintain a "reconciler loop" as such. It's entirely up to the operator/controller to configure watches/caches against the API and trigger reconciliations based on that.

marcoandredinis replied:
Updated accordingly
I removed the `reconciler loop` box, I think it's ok to label the entire system as a single box and only expose the necessary details regarding interactions it has from the user's and developer's point of view.

strideynet commented on operator/main.go:
```diff
@@ -0,0 +1,165 @@
+	}
+
+	// Install CRDs
+	if err := mgr.Add(manager.RunnableFunc(func(ctx context.Context) error {
```
Have we confirmed that this isn't racy with the Controller watches/caches being established ?

marcoandredinis replied:
I tried this multiple times and didn't found an issue, but can't guarantee it won't :sweat_smile: 

I added a `WaitForCacheSync` before installing the CRDs, it should prevent races in case they are possible. Better be safe

strideynet replied:
Ah yeah, I should have expanded, but my concern was that I couldn't recall how the kube api handles you trying to establish a watch on a resource that hasn't yet been defined. In previous operators I have built, we have had the user manually deploy the CRDs prior to deploying the operator, I definitely prefer this model of the operator deploying it's own CRDs though.

strideynet replied:
Perfect, those logs confirm that we handle this situation. I think the retry negates the need for some additional complexity here.

strideynet reviewed (commented):
This looks good, I've just got a few queries/nits that would be good to confirm

marcoandredinis commented on lib/kube/proxy/roundtrip.go:
```diff
@@ -170,13 +166,9 @@ func (s *SpdyRoundTripper) RoundTrip(req *http.Request) (*http.Response, error)
 	)
 
-	if s.followRedirects {
-		conn, rawResponse, err = utilnet.ConnectWithRedirects(req.Method, req.URL, header, req.Body, s, false)
```
That's my assumption as well
The lib we are using removed them:
https://github.com/kubernetes/apimachinery/commit/30f68c403333a6bf4abe84c02d934e82836a08fc
Unfortunately they didn't provide any reasoning to removing them, and my understanding of SPDY or where we use this is limited (besides, also being related to kubernetes)

If we feel like this should be reverted, we can fork the upstream and revert their commit and revert the 2nd commit of this PR.
I would rather not do it, unless we really need to.

marcoandredinis replied:
Looking at those PRs, this feature seems to be a bit of an edge case
It was deprecated in 1.18 and has no effect in 1.20 afterwards

Do you know what's our lowest supported version?
I'm having troubles starting `minikube` using version 1.17 or 1.18 (the docker version is greater than supported)

tigrato replied:
I am happy if it's working with 1.20

marcoandredinis commented on operator/apis/resources/v2/user_types.go:
```diff
@@ -0,0 +1,98 @@
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// UserSpec defines the desired state of User
+type UserSpec types.UserSpecV2
```
> I do wonder if reusing the types.*Spec for the Kubernetes resource specs is ever going to cause us pain here. 

It will :sweat_smile: 
We have another project where we re-use the entire `types.*Spec`: Terraform Provider for Teleport.
https://github.com/gravitational/teleport-plugins/tree/master/terraform

It's not always pretty but works 99% of the time and saves us tons of development time.
When it doesn't work, we usually need to implement some custom type or some exception, but so far is proving to be worthwhile.

I do agree with you, if it stops working we can always be explicit about the spec and matching JSON Keys.

strideynet replied:
Sounds good, wasn't aware we had prior art on this. I haven't had a play with our Terraform provider yet :')

marcoandredinis commented on operator/apis/resources/v2/user_types.go:
```diff
@@ -0,0 +1,98 @@
+//+kubebuilder:subresource:status
+
+// User is the Schema for the users API
+type User struct {
```
Yeah, I'm not sure it would work out without some core changes

There's also another thing: we are already `Check` and `SetDefaults` on the Teleport API
Having two places calling the same code in the same request doesn't seem the right thing to do given those applications are meant to be side by side (in a sidecar architecture)

strideynet replied:
Yeah, that's fair. I was just wondering about feedback to the user. If they are mainly configuring it via k8s custom resources, it might be nice if the defaults were reflected back to them after submitting the manifest. Let's revisit this at a later date though, I think it's fine as it is.

marcoandredinis commented on operator/controllers/resources/role_controller_test.go:
```diff
@@ -0,0 +1,186 @@
+)
+
+func TestRoleCreation(t *testing.T) {
+	ctx := context.Background()
```
Yes, thank you :pray:

marcoandredinis replied:
Sorry, I ended up reverting this change
The tests were passing locally, but failing in the CICD

tigrato commented on operator/config/rbac/auth_proxy_role.yaml:
```diff
@@ -0,0 +1,17 @@
+- apiGroups:
+  - authorization.k8s.io
+  resources:
+  - subjectaccessreviews
```
Do you know why it requires `subjectaccessreviews` without any scope?

marcoandredinis replied:
Please see this comment
https://github.com/gravitational/teleport/pull/13331#issuecomment-1161460367

tigrato commented on operator/config/rbac/auth_proxy_client_clusterrole.yaml:
```diff
@@ -0,0 +1,9 @@
+  name: metrics-reader
+rules:
+- nonResourceURLs:
+  - "/metrics"
```
which metrics are we expected to scrape?

tigrato replied:
Does the agent expose it to external endpoints, i.e., prometheus or does it consume any metric? If it just exposes them it does not requires get access to any `/metrics`

marcoandredinis replied:
I don't think the agent consumes any of those.

We only expose them.

marcoandredinis replied:
Please see this comment
https://github.com/gravitational/teleport/pull/13331#issuecomment-1161460367

marcoandredinis commented:
Hey @strideynet thank you for looking into this
I know this was a long PR, so double thank you :pray:

tigrato commented on operator/controllers/resources/role_controller.go:
```diff
@@ -0,0 +1,88 @@
+	if err != nil && !trace.IsNotFound(err) {
+		return trace.Wrap(err)
+	}
+	if trace.IsNotFound(err) {
```
```suggestion
	if trace.IsNotFound(err) && !hasOriginLabel(obj) {
	  return addOriginLabel(ctx, r.Client, obj)
	}
```

tigrato reviewed (commented):
I am generally ok with the approach but check if by removing the `proxy` forwarder we do not break Kubernetes access for older clusters.

Check my RBAC questions as well

marcoandredinis commented:
Trying to clarify a bit the contents of the `operator/config` folder:

Everything inside the `operator/config` folder was auto generated on project's bootstrap (first commit)
Well, except for the CRDs for Roles and Users: `operator/config/crd/bases/resources.teleport.dev_*.yaml`, which were generated by `crdgen` and after creating the actual User and Role Specs.

We only include those `operator/config/crd/bases/*` files in the `teleport-operator` binary, and the only output of this project is container image which only has that same binary.
This means the operator will not be aware of any other file.

I'm going to remove all those unused files

marcoandredinis commented:
@tigrato @hugoShaka @r0mant  @strideynet 
Me and Hugo revisited the CRD flow and I think we have a better solution (props to Hugo)
It's closer to what Tiago was suggesting

The idea is to create a subchart that installs the CRDs using templating and not the `crd` folder

I created this Draft PR that shows how we can accomplish it
https://github.com/gravitational/teleport/pull/13754

What do you think?
If the approach looks good, I'll remove the CRD installation from this PR and, when this one is reviewed and approved, open up a new one with the Chart and operator logic.

strideynet commented:
@marcoandredinis So essentially, there will be a first time install of the CRDs via Helm, but then we handle upgrading them (since Helm can't handle that), via the operator ?

marcoandredinis commented:
> @marcoandredinis So essentially, there will be a first time install of the CRDs via Helm, but then we handle upgrading them (since Helm can't handle that), via the operator ?
> 
> If so, I think that's an approach I can support.

No, we would upgrade the CRDs in the subchart, and when applying the new subchart version it would trigger the CRD upgrade in the cluster.
This works for new non-required fields, which is ideal.

Ultimately, If we add a required field, we can ignore its required prop, and just sent it to Teleport, which returns an error and we return that error in the reconciler

strideynet commented:
> > @marcoandredinis So essentially, there will be a first time install of the CRDs via Helm, but then we handle upgrading them (since Helm can't handle that), via the operator ?
> > If so, I think that's an approach I can support.
> 
> No, we would upgrade the CRDs in the subchart, and when applying the new subchart version it would trigger the CRD upgrade in the cluster. This works for new non-required fields, which is ideal.
> 
> Ultimately, If we add a required field, we can ignore its required prop, and just sent it to Teleport, which returns an error and we return that error in the reconciler

Ah ok, a bit more complex. It still sounds relatively sound as a strategy though. re: adding new required fields, it's relatively awkward to handle that in k8s whichever way you do it. They've always tried to just encourage having relatively stable resource schemas within major versions of the API.

marcoandredinis commented:
> * is it possible for Helm upgrade strategy to delete and recreate the crd? If a deletion happens either because we change the name or the crd filename, does K8S delete resources for that CRD (I have no idea if it's the case)?  If this happens, it might be no-go for that solution.

Adding new fields does not recreate the CRD, neither deletes any Custom Resources the cluster already has.
Changing the CRD name will actually create another CRD because the name is the primary key for the CRDs
Changing the CRD filename has no impact

Here's the test I did:
- install chart with CRDs
- created a user CR and a role CR
- edited the subchart:
  - renamed the user crd file to `thing.yaml` and added a new field
  - renamed the role CRD (kept the same file name) from `role` to `roleabc`
- upgraded the chart with the new version
- the CRD for the user had a new field and the user CR didn't change
- the CRD for the role didn't change and the role CR didn't change
- the CRD for the roleabc was created

> * When installing crds from the operator, can we inject custom code? I am thinking about upgrade strategies where we run code that reads resources and updates them for the new CRD spec.

I would say it's possible.
We have access to the HELM hooks to inject any custom code we want

marcoandredinis commented:
Created this Issue as a followup for this PR regarding the update of CRDs
https://github.com/gravitational/teleport/issues/13849

Joerger reviewed (commented):
I still need to look through `/controllers` and a few other files, but LGTM overall. Just some nits on the readme again.

marcoandredinis commented on go.mod:
```diff
@@ -24,7 +24,7 @@ require (
 	github.com/beevik/etree v1.1.0
 	github.com/cloudflare/cfssl v0.0.0-20190726000631-633726f6bcb7
-	github.com/coreos/go-oidc v0.0.0
+	github.com/coreos/go-oidc v2.1.0+incompatible
```
It's a dependecy from k8s.io/apiserver:

```
$ go mod graph | grep oidc
github.com/gravitational/teleport github.com/coreos/go-oidc@v2.1.0+incompatible
k8s.io/apiserver@v0.24.0 github.com/coreos/go-oidc@v2.1.0+incompatible
```

marcoandredinis commented on version.go:
```diff
@@ -3,7 +3,7 @@
 
 const (
-	Version = "11.0.0-dev"
+	Version = "10.0.0-dev.marco.3"
```
Sorry :sweat_smile:

marcoandredinis commented on operator/Makefile:
```diff
@@ -0,0 +1,260 @@
+# To re-generate a bundle for another specific version without changing the standard setup, you can:
+# - use the VERSION as arg of the bundle target (e.g make bundle VERSION=0.0.2)
+# - use environment variables to overwrite this value (e.g export VERSION=0.0.2)
+VERSION ?= 0.0.1
```
It will inherit the root's Makefile, so it should always receive the correct value.

I'm ok changing this to have the correct version
However, we might need to add an extra step when creating a new release to update this.
I don't see a target in the root's Makefile that updates the version, so we would probably need a new target

Adding an extra manual step (either calling a new target or manually changing the `/operator/Makfile`) doesn't seem to be worth, but I'm fine with either option if you feel strong about this :+1:

marcoandredinis commented on operator/README.md:
```diff
@@ -0,0 +1,151 @@
+
+`$ kubcetl apply -f roles.yaml`
+
+And now check if the role was created in Teleport and K8S (`teleport-cluster` namespace).
```
Did you change the default namespace to be the one we just created?
```bash
kubectl config set-context --current --namespace teleport-cluster
```

Btw, if you want to see a more up to date version use the dev build `11.0.0-marco.3`

```bash
helm upgrade --install --create-namespace -n teleport-cluster \
	--set clusterName=teleport-cluster.teleport-cluster.svc.cluster.local \
	--set teleportVersionOverride="11.0.0-marco.3" \
	--set operator=true \
	teleport-cluster ${TELEPORT_PROJECT}/examples/chart/teleport-cluster
```

marcoandredinis commented on operator/README.md:
```diff
@@ -0,0 +1,151 @@
+
+First, let's create a tunnel using minikube: `minikube tunnel` (this command runs is foreground, open another terminal for the remaining commands).
+
+Now, let's create a new Teleport User and login in the web UI:
```
`teleoperator` is the user we create to navigate the Teleport's webui
We can just create another user

This can also be from a previous version we had (tagged as 9.3.5)
Can you please test the same using the dev build?

```bash
minikube delete; minikube start;
export TELEPORT_PROJECT=<github.com/gravitation/teleport checked out at `marco/plugins-teleport-operator-charts`>
helm upgrade --install --create-namespace -n teleport-cluster \
	--set clusterName=teleport-cluster.teleport-cluster.svc.cluster.local \
	--set teleportVersionOverride="11.0.0-marco.3" \
	--set operator=true \
	teleport-cluster ${TELEPORT_PROJECT}/examples/chart/teleport-cluster
```
(You don't need to run `minikube delete` afterwards, this is just to make sure we clean the state of previous builds that might have created the wrong sidecar user)

Joerger replied:
Yep, looks like running `minikube delte` and using the newer build fixes the issue, thanks!

marcoandredinis commented on operator/crdgen/ignored.go:
```diff
@@ -0,0 +1,23 @@
+import "github.com/Clever/go-utils/stringset"
+
+var ignoredFields = map[string]stringset.StringSet{
+	"UserSpecV2": stringset.New("LocalAuth", "Expires", "CreatedBy", "Status"),
```
These fields are read-only and can not be set from the Teleport's client
CRDs don's support `readonly` fields (it's scheduled to land in 1.25 - https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#transition-rules)

There are probably more readonly fields that are not here, only doing an exhaustive search we would be able to find all of them

We can remove this logic and always send every field, and let Teleport API return the appropriate error.

marcoandredinis replied:
Added a comment

marcoandredinis commented on operator/README.md:
```diff
@@ -0,0 +1,159 @@
+### Requirementes
+
+#### K8S cluster
+You can use `minikube` if you don't have a cluster yet.
```
The goal is to give the reader an easy way to test this out, without messing with a real cluster.

`minikube` gives out that

What do you suggest here?
I think suggesting the reader/developer to setup a "real cluster" or to try this out in an existing one is a bit too much
We suggest `minikube` in other parts of our documentation, so I think we should be good here

tigrato commented on operator/controllers/resources/reconciler.go:
```diff
@@ -0,0 +1,123 @@
+	}
+
+	for k := range obj.GetLabels() {
+		if k == types.OriginLabel {
```
labels are a map so the whole function can be replaced with:
```go
	_,ok :=obj.GetLabels()[types.OriginLabel]
	return ok
```

You don't even need to check for nillness

tigrato commented on operator/controllers/resources/role_controller.go:
```diff
@@ -0,0 +1,87 @@
+	if err != nil && !trace.IsNotFound(err) {
+		return trace.Wrap(err)
+	}
+	if trace.IsNotFound(err) && !hasOriginLabel(obj) {
```
change the `addOriginLabel` to reflect that it's affecting the K8S object, not the Teleport object.

tigrato reviewed (approved):
Check my comments but LGTM

marcoandredinis commented on operator/controllers/resources/reconciler.go:
```diff
@@ -0,0 +1,123 @@
+	// Create or update
+	log.Info("upsert object in Teleport")
+	if err := r.UpsertExternal(ctx, obj); err != nil {
+		return ctrl.Result{}, trace.Wrap(err)
```
Returning an error should re-restart (re-enqueue) the event reconciliation
https://github.com/kubernetes-sigs/controller-runtime/blob/ffd9ec8768b7d2e82aee6b0c76d481a5bdda4075/pkg/reconcile/reconcile.go#L91-L92

marcoandredinis commented:
Thank you very much for your review @tigrato 
Sorry for the painfully long change set :pray:

Joerger commented on operator/controllers/resources/reconciler.go:
```diff
@@ -0,0 +1,142 @@
+	if isMarkedToBeDeleted {
+		if hasDeletionFinalizer {
+			log.Info("deleting object in Teleport")
+			if err := r.DeleteExternal(ctx, obj); err != nil {
```
Should this handle not found error, in the case where the resource was deleted directly through Teleport rather than the operator? Right now this leads to:

```
1.6565323828787718e+09	INFO	deleting object in Teleport	{"controller": "user", "controllerGroup": "resources.teleport.dev", "controllerKind": "User", "role": {"name":"myrole","namespace":"teleport-cluster"}, "namespace": "teleport-cluster", "name": "myrole", "reconcileID": "846a414c-d2b9-457a-85c7-61c12fe713c4", "namespacedname": "teleport-cluster/myrole"}
1.6565323828816507e+09	ERROR	Reconciler error	{"controller": "user", "controllerGroup": "resources.teleport.dev", "controllerKind": "User", "role": {"name":"myrole","namespace":"teleport-cluster"}, "namespace": "teleport-cluster", "name": "myrole", "reconcileID": "846a414c-d2b9-457a-85c7-61c12fe713c4", "error": "role \"myrole\" is not found"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.1/pkg/internal/controller/controller.go:273
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.1/pkg/internal/controller/controller.go:234
```

Can you add a test case for this scenario as well?

hugoShaka replied:
Oh, great point. It feels weird to have the operator not reconcile when something happens to one of its managed resources, this is a common pain point when reconciling things not in Kubernetes.

Ideally it should be doing a WATCH to get updates from resources in Teleport but this can be overly complex to do. Maybe a Tick every X minute looking for changes would be enough (but this could drastically increase DB usage). A janky fix workaround could also be reducing the resync period from its default 10 hours to something shorter like 1 hour.

marcoandredinis replied:
I committed a failing test and a fix for it
When trying to delete but it errors with "NotFound" we continue as if it was not an error

Regarding the sync from Teleport into Kubernetes' CR, we decided to not implement that on our first iteration
We only support from Kubernetes to Teleport

Sorry, this is not written anywhere, afaik

hugoShaka replied:
In this case I suggest reducing the resync period (maybe have this configurable?), it will ensure we converge back way faster if a drift happens

hugoShaka commented on operator/crdgen/testdata/golden/resources.teleport.dev_users.yaml:
```diff
@@ -0,0 +1,120 @@
+spec:
+  group: resources.teleport.dev
+  names:
+    kind: User
+    listKind: UserList
+    plural: users
+    singular: user
```
The terms Users and Roles are confusing in a Kubernetes context.

This creates weird scenarios like
```
kubectl api-resources | grep roles
NAME             SHORTNAMES   APIVERSION                             NAMESPACED   KIND
clusterroles                  rbac.authorization.k8s.io/v1           false        ClusterRole
roles                         rbac.authorization.k8s.io/v1           true         Role
roles                         resources.teleport.dev/v5              true         Role
```

In case of conflict the Resources brought by CRDs will loose over the core Resources, but this means our resources are not reachable through their singular name.

```
shaka> kubectl get roles -n teleport-cluster-crds
NAME                        CREATED AT
teleport-cluster-crds-psp   2022-06-29T19:41:49Z

shaka> kubectl get roles.resources -n teleport-cluster-crds
No resources found in teleport-cluster-crds namespace.

shaka> kubectl get roles.resources.teleport -n teleport-cluster-crds
No resources found in teleport-cluster-crds namespace.

shaka> kubectl get roles.resources.teleport.dev -n teleport-cluster-crds
No resources found in teleport-cluster-crds namespace.
```

When there are no conflicts this still feels weird, this command really doesn't look as it is doing a teleport-specific thing.

```
shaka> kubectl get users -n teleport-cluster-crds
No resources found in teleport-cluster-crds namespace.
```

While the user can still specify the full GVR it is not a great user experience.

What do you think about:
- Using different Kinds ? `TeleportUser`, `TeleportRole` for example. Maybe with the shortnames `tpuser`, `tprole`
- Keeping the Kinds but changing the resource? GVK would still be `resources.teleport.dev/v5/Role` but GVR would be `resources.teleport.dev/v5/teleportRole`. I'm not sure this is a good Idea as even advanced Kubernetes users tend to mix Kinds and Resources
-  Keeping everything as-is and using the shortname to have a usable way to call the resource

marcoandredinis replied:
> `shaka> kubectl get roles.resources.teleport.dev -n teleport-cluster-crds`

This one should return your Teleport Roles
Did you have any?

I tested by creating a role and querying it with:
```bash
$ k get roles.resources.teleport.dev
NAME      AGE
myrolex   11m
```
It outputs the same for `k get roles.resources` and `k get roles.resources.teleport`

I'm ok prepending `Teleport` or `teleport`
Adding `tp` seems a bit weird because we don't use that abbreviation anywhere (do we?)

I'm also ok leaving it is

@r0mant Do you think this should be a product decision?

hugoShaka replied:
As discussed with @r0mant this will likely be a product decision. Here are the examples for each solution:

## Solution 1: Shortname conflict but Kind doesn't

Prefix all our `Kind` with `Teleport` and provide shortnames without the prefix. Shortnames might conflict with other resources but api-resource and Kind won't.

Commands example
```
kubectl api-resources | grep roles
NAME             SHORTNAMES     APIVERSION                             NAMESPACED   KIND
clusterroles                    rbac.authorization.k8s.io/v1           false        ClusterRole
roles                           rbac.authorization.k8s.io/v1           true         Role
teleportroles    role           resources.teleport.dev/v5              true         TeleportRole
```

Manifest example
```
apiVersion: resources.teleport.dev/v5
kind: TeleportRole
metadata:
  name: admin
spec: {}
``` 

## Solution 2: Kind conflicts but shortnames does't

Keep our conflicting kinds and use shortnames to create unambiguous names.

Commands example
```
kubectl api-resources | grep roles
NAME             SHORTNAMES     APIVERSION                             NAMESPACED   KIND
clusterroles                    rbac.authorization.k8s.io/v1           false        ClusterRole
roles                           rbac.authorization.k8s.io/v1           true         Role
roles            teleportrole   resources.teleport.dev/v5              true         Role
```

Manifest example
```
apiVersion: resources.teleport.dev/v5
kind: Role
metadata:
  name: admin
spec: {}
``` 

## Solution 3: Differentiating GVK and GVR

Usually we see a 1-1 relation between the GVR (API resource name + Group + Version} and the GVK (Kind + Group + Version), but Kubernetes allow the same Kind to be exposed at multiple places in the API. This is used to move seamlessly stuff from one api to another, e.g. during the `Deployment` migration from `extensions/v1` to `apps/v1`. I have not heard of a Kind being exposed with a different resource name. I feel this trick will be confusing for some users.

```
kubectl api-resources | grep roles
NAME             SHORTNAMES     APIVERSION                             NAMESPACED   KIND
clusterroles                    rbac.authorization.k8s.io/v1           false        ClusterRole
roles                           rbac.authorization.k8s.io/v1           true         Role
teleportroles    role           resources.teleport.dev/v5              true         Role
```

Manifest example
```
apiVersion: resources.teleport.dev/v5
kind: Role
metadata:
  name: admin
spec: {}
```

hugoShaka replied:
I strongly suggest solution 1 because its Kinds are 100% unambiguous.

As a user dealing with a manifest of a `TeleportUser` I know it will create a user in Teleport, and not in Kubernetes, AWS or some other system.

When we'll add other resources like `Application`, writing a `TeleportApplication` will feel like I'm configuring application access in teleport. However writing an `Application` resource in Kubernetes might feel like the resource is deploying the application itself.

hugoShaka replied:
Product agreed to switch to `TeleportUser` and `TeleportRole`

Joerger commented on operator/controllers/resources/user_controller.go:
```diff
@@ -0,0 +1,91 @@
+		return trace.Wrap(r.TeleportClient.CreateUser(ctx, teleportResource))
+	}
+
+	return trace.Wrap(r.TeleportClient.UpdateUser(ctx, teleportResource))
```
If the user was created via Teleport and does not have `labels: teleport.dev/origin: kubernetes`, what should the operator do?

Currently it is overwriting the user still, including the ignored fields, and still does not add the origin label:
```
kind: user
metadata:
  id: 1656534039153017096
  name: dev
spec:
  created_by:
    time: "2022-06-29T20:20:39.152810861Z"
    user:
      name: 340fb098-21ce-4c31-ae6d-3dd0314c25d5.teleport-cluster.teleport-cluster.svc.cluster.local
  expires: "0001-01-01T00:00:00Z"
  roles:
  - myrole
  status:
    is_locked: false
    lock_expires: "0001-01-01T00:00:00Z"
    locked_time: "0001-01-01T00:00:00Z"
    recovery_attempt_lock_expires: "0001-01-01T00:00:00Z"
  traits:
    aws_role_arns: null
    db_names: null
    db_users: null
    kubernetes_groups: null
    kubernetes_users: null
    logins:
    - bjoerger
    windows_logins: null
version: v2
```
-->
```
kind: user
metadata:
  id: 1656534245019790643
  name: dev
spec:
  created_by:
    time: "0001-01-01T00:00:00Z"
    user:
      name: ""
  expires: "0001-01-01T00:00:00Z"
  roles:
  - myrole
  status:
    is_locked: false
    lock_expires: "0001-01-01T00:00:00Z"
    locked_time: "0001-01-01T00:00:00Z"
    recovery_attempt_lock_expires: "0001-01-01T00:00:00Z"
  traits:
    logins:
    - bjoerger
version: v2
```

hugoShaka commented on operator/controllers/resources/user_controller.go:
```diff
@@ -0,0 +1,91 @@
+		Complete(r)
+}
+
+func (r *UserReconciler) Delete(ctx context.Context, obj kclient.Object) error {
+	return r.TeleportClient.DeleteUser(ctx, obj.GetName())
+}
```
The delete does not check the unicity of the link between the CR and the teleport resource. The following scenario demonstrates a drift between the real state and the desired state.

```
$ cat user.yaml
apiVersion: resources.teleport.dev/v2
kind: User
metadata:
  name: eva
spec:
  roles:
    - yolo
  traits:
    logins:
      - root

$ kubectl create -f user.yaml -n ns1
# the user is now present in Kubernetes and Teleport

$ kubectl create -f user.yaml -n ns2
# the user is now present twice in Kubernetes and once in Teleport

$ kubectl delete -f user.yaml -n ns1
# the user is now present once in Kubernetes and is not in teleport
```

Such drift will stay until a new even happen on the user, the controller is restarted, or until a full resync (every 10 hours).

This raises 2 questions:
- Proof of ownership from the operator (if the user was here before the operator, we create and delete it, shoult it be deleted from teleport as well, @Joerger  is also raising the question here: https://github.com/gravitational/teleport/pull/13331/files#r910388983)
- Proof of 1-1 mapping between the kube ressource and the teleport resource. What happens if there are multiple operators fighting for the same resource? Multiple resources trying to represent the same user?

I think it might be saner for the operator to:
- label the namespace the resource is coming from (name + namespace should be a unique idenfier in a Kube cluster)
- label the resources with the release name, or at least a unique key we give to the operator (ensure we can ensure unicity across multiple Kubernetes clusters, `external-dns` does this and it saved my life a couple of times)
- refuse to interact with a teleport resources incorrectly labeled

It means users wanting to "adopt" existing Teleport resources in Kubernetes will have to label them accordingly. This is an extra step but might be safer.

This would also allow us to be idempotent and clean resources on startup even if a finalizer failed (if the Kubernetes cluster was down/destroyed for some reason) and we have no trace of the Kube CR.

marcoandredinis replied:
The first part should be solved, because we added a restriction to the namespaces the controller is watching.
So, in this case `kubectl create -f user.yaml -n ns2` would either not work or create the resource in another Teleport Cluster

The other topic seems a lot more tricky
What should we do when states diverge? either because it already existed
a) merge the properties
b) only update or create **if the resource has our origin label** and return an error otherwise
c) always update the resource, no matter what

Right now we have c)
I kinda of like option b) (proof of ownership or ownership handoff)

@Joerger I think your comment above is about this issue, so let's discuss it in this thread
https://github.com/gravitational/teleport/pull/13331#discussion_r910388983
If not, please let me know

@r0mant any strong opinion on this?

Joerger replied:
I don't have a strong preference either way, but if we go with the current approach (c), then an update should always result in the origin label getting added, and should not update ignored fields I think? In my comment it shows how `created_by` gets set to a zero value in this case, but in the normal operator create -> operator update case, the ignored fields like `created_by` get properly ignored.

hugoShaka replied:
I gave a try in the PR https://github.com/gravitational/teleport/pull/14099

- Following a discussion with @r0mant: I implemented solution b
- The PR adds a user feedback through the use of status.conditions
- It also adds custom resource definition status generation (with a couple of ugly fixes to locate packages, I'll try to revamp this later) from go types in `crdgen`

> Also, note that currently (as far as I can tell) there is no way to create a new user through the operator which can be logged in with. You are forced to `kubectl exec -- tctl users add user --roles=roles` to get the invite token, and then if you want you can update the user afterwards with the operator. Currently this flow leads to the problematic case described above.

Users can (and should?) use UI/CLI to get reset links on resources created by the operator, as for the Terraform provider. Either by doing `tctl users reset myuser` or clicking on the `Reset password` button on the Users UI tab.

tigrato commented on operator/main.go:
```diff
@@ -0,0 +1,143 @@
+
+	ctrl.SetLogger(zap.New(zap.UseFlagOptions(&opts)))
+
+	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
```
@hugoShaka I think that the biggest issue is not being able to create resources in multiple namespaces since you can always limit the user scope via RBAC. 
The biggest issue is that resources in other namespaces can replace each other. If user A creates a user in namespace A, another user can create the same user in namespace B  and it will replace the initial spec.

marcoandredinis replied:
Thank you
I'm now restricting the namespace we subscribe

marcoandredinis commented on operator/main.go:
```diff
@@ -0,0 +1,143 @@
+
+	retry, err := utils.NewLinear(utils.LinearConfig{
+		Step: 100 * time.Millisecond,
+		Max:  time.Second,
```
I guess it depends on the configuration
Doing a small test locally by starting the Auth, SSH and Proxy services, it took around 4 seconds

> It's a bit confusing to have non-zero restart counters on the container by default.

It doesn't restart unless it takes too long to connect create the client.
Did you get a restart?
Not sure you looked into the `utils.NewLinear`, but that `Max` value can be confusing
`Max` is to max sleep duration between retries.
So, it starts by sleeping 100ms, then 200ms until it reaches the Max value (in this case 1000ms)

If you have a suggestion for other values here, I'm fine with changing them :+1:

hugoShaka replied:
Oh my bad, I don't understand yet how `utils.NewLinear` works.

marcoandredinis commented on operator/crdgen/schemagen.go:
```diff
@@ -0,0 +1,290 @@
+						},
+						"metadata": {Type: "object"},
+						"spec":     schema.JSONSchemaProps,
+						"status": {
+							Type:        "object",
+							Description: fmt.Sprintf("%s resource status", root.kind),
+							Properties: map[string]apiextv1.JSONSchemaProps{
+								"lastError": {
+									Type: "object",
+									Properties: map[string]apiextv1.JSONSchemaProps{
+										"kind": {
+											Type:        "string",
+											Description: "Error kind",
+										},
+										"text": {
+											Type:        "string",
+											Description: "Error text",
+										},
+									},
+								},
+							},
```
Thank you! I overlooked this one

I'll remove the status generation in the `operator/crdgen/schemagen.org`
A new `make -C operator manifests` should remove `status` in the CRD

hugoShaka replied:
Should be resolved with https://github.com/gravitational/teleport/pull/14099

hugoShaka commented on operator/main.go:
```diff
@@ -0,0 +1,143 @@
+		Port:                   9443,
+		HealthProbeBindAddress: probeAddr,
+		LeaderElection:         true,
+		LeaderElectionID:       "431e83f4.teleport.dev",
```
Makes sense to expose this as a flag. What `431e83f4` stands for?

If the identifier is built from `helm-release+namespace` I would allow to run the operator multiple times, even in the same namespace.

marcoandredinis replied:
It was randomly generated by the `operator-sdk` initialization
No problem in dropping it :+1:

hugoShaka commented on operator/Makefile:
```diff
@@ -0,0 +1,260 @@
+ENVTEST_K8S_VERSION = 1.23
+
+# Get the currently used golang install path (in GOPATH/bin, unless GOBIN is set)
+# TODO(marco): do we need this?
+#ifeq (,$(shell go env GOBIN))
+#GOBIN=$(shell go env GOPATH)/bin
+#else
+#GOBIN=$(shell go env GOBIN)
+#endif
```
Is this TODO still relevant?

marcoandredinis replied:
We can remove it :+1:

hugoShaka commented on operator/Makefile:
```diff
@@ -0,0 +1,260 @@
+
+KUSTOMIZE_INSTALL_SCRIPT ?= "https://raw.githubusercontent.com/kubernetes-sigs/kustomize/master/hack/install_kustomize.sh"
+.PHONY: kustomize
+kustomize: $(KUSTOMIZE) ## Download kustomize locally if necessary.
+$(KUSTOMIZE): $(LOCALBIN)
+	curl -s $(KUSTOMIZE_INSTALL_SCRIPT) | bash -s -- $(subst v,,$(KUSTOMIZE_VERSION)) $(LOCALBIN)
+
```
If I understand correctly we don't plan to use Kustomize anymore. If this is the case this should be removed.

marcoandredinis replied:
True
It's part of the project's bootstrap features we are not using

hugoShaka replied:
removed

hugoShaka commented:
Hello reviewers :) , @tigrato @Joerger @strideynet @marcoandredinis 

I am continuing marco's PR and did a couple of major changes. Could you review it again? (I'm sorry, I know this PR is a monster).

Main areas that were changed and you might want to focus on are:
- Helm charts changes in `teleport-cluster` to integrate the operator + new `teleport-operator` subchart
- Kinds were renamed from `Role` and `User` to `TeleportRole` and `TeleportUser`
- The operator no longer put the ownership label on the Kubernetes resources (label is only added on the Teleport resources)
- The operator will now refuse to interact with an existing Teleport resource that doesn't have the label
- If a reconciliation fails it will be reflected on the status (currently, only if resource is not owned, I will add generic error catching in a subsequent PR)
- The crd generator populates `Status` from go types

Joerger reviewed (commented):
The README is out of date, are we planning on replacing the README with docs?

marcoandredinis commented on examples/chart/teleport-cluster/values.yaml:
```diff
@@ -54,6 +54,20 @@ acmeURI: ""
 # kubectl -n ${TELEPORT_NAMESPACE?} create secret generic license --from-file=/path/to/downloaded/license.pem
 enterprise: false
 
+# Configuration of the optional Teleport operator
+operator:
```
I prefer the separation here
However, I think we are missing the `image` here (it currently exists as `operatorImage`)

marcoandredinis commented on examples/chart/teleport-cluster/templates/deployment.yaml:
```diff
@@ -205,6 +205,33 @@ spec:
+            port: 8081
+          initialDelaySeconds: 5
+          periodSeconds: 10
+{{- if .Values.resources }}
+        resources:
+  {{- toYaml .Values.resources | nindent 10 }}
+{{- end }}
```
I think this one should be`.Values.operator.resources` instead of `.Values.resources`

marcoandredinis commented on operator/README.md:
```diff
@@ -0,0 +1,158 @@
+helm upgrade --install --create-namespace -n teleport-cluster \
+	--set clusterName=teleport-cluster.teleport-cluster.svc.cluster.local \
+	--set teleportVersionOverride="11.0.0-dev" \
+	--set operator=true \
```
`operator.enabled=true`

hugoShaka replied:
Fixed

marcoandredinis commented on operator/config/crd/bases/resources.teleport.dev_roles.yaml:
```diff
@@ -0,0 +1,978 @@
+apiVersion: apiextensions.k8s.io/v1
```
I think we can remove this file
They are copied into the subchart

The same goes to the Teleport User CRD

hugoShaka replied:
Related to the discussion: https://github.com/gravitational/teleport/pull/13331#discussion_r917608473

marcoandredinis commented on operator/controllers/resources/utils.go:
```diff
@@ -0,0 +1,58 @@
+// checkOwnership takes an existing resource and validates the operator owns it.
+// It returns an ownership condition and an error if the resource is not owned by the operator
+func checkOwnership(existingResource types.Resource) (metav1.Condition, error) {
+	if existingResource != nil {
+		if !isResourceOriginKubernetes(existingResource) {
```
:nail_care: 
I think we could decrease this function's max indentation by checking first if `existingResource == nil`

marcoandredinis commented on operator/controllers/resources/suite_test.go:
```diff
@@ -0,0 +1,180 @@
+
+func startKubernetesOperator(t *testing.T, teleportClient *client.Client) kclient.Client {
+	testEnv := &envtest.Environment{
+		CRDDirectoryPaths:     []string{filepath.Join("..", "..", "config", "crd", "bases")},
```
Changing the CRD output path from `operator/config/crd/bases/` to `examples/chart/teleport-cluster/charts/teleport-operator/templates` means this will stop working

We have to either write to the two places or change this path to the subchart location
I prefer the latter

marcoandredinis commented on operator/sidecar/sidecar.go:
```diff
@@ -0,0 +1,250 @@
+		return trace.Wrap(err)
+	}
+
+	reqExpiry := time.Now().UTC().Add(1 * time.Hour)
```
Sorry, I think we have to change this
It's basically creating a certificate that allows our client to access the API for 1 hour, which will not be enough
We have to either increase the expiration time or the client must be able to re-create a certificate when the one we have expires

hugoShaka replied:
This has been addressed in https://github.com/gravitational/teleport/pull/13331/commits/0fa831199553e0dec95075bf4ef417ea8f934a21

This is a pretty big change. tl;dr
- Local auth tsh-style connection happens on startup to create roles
- We wait for leader election
- When leader it registers a new teleport bot (bot currently don't support multiple tokens on the same bot, we have de delete/recreate)
- When teleport bort is registered we start a `tbot`, it renews its own cert and the client one
- For each request we create a new client using the certs from the tbot memory. This is a bit ineffective, at some point if tbot exposes a renewal channel we can listen on this should be improved.

tigrato commented on operator/controllers/resources/utils.go:
```diff
@@ -0,0 +1,58 @@
+
+// isResourceOriginKubernetes reads a teleport resource metadata, searches for the origin label and checks its
+// value is kubernetes.
+func isResourceOriginKubernetes(resource types.Resource) bool {
+	metadata := resource.GetMetadata()
+	if label, ok := metadata.Labels[types.OriginLabel]; ok {
+		return label == types.OriginKubernetes
+	}
+	return false
+}
```
```suggestion
func isResourceOriginKubernetes(resource types.Resource) bool {
	label := resource.GetMetadata().Labels[types.OriginLabel]
	return label == types.OriginKubernetes
}
```

tigrato commented on operator/controllers/resources/role_controller.go:
```diff
@@ -0,0 +1,108 @@
+	}
+
+	if ownershipErr != nil {
+		return trace.Wrap(ownershipErr)
```
If not owned by the operator and  we return an error, won't it requeue the operation forever?

hugoShaka replied:
Yes, but I'm not sure if the alternative of not caring anymore is saner. If the user edits the resource to allow adoption it won't be reconciled. Maybe this could be handled with another status.condition.

hugoShaka commented on examples/chart/teleport-kube-agent/Chart.yaml:
```diff
@@ -1,7 +1,9 @@
 apiVersion: v2
-version: "11.0.0-dev"
-appVersion: "11.0.0-dev"
+version: *version
```
I swicthed to anchors for the `teleport-cluster` chart and its subchart dependency.

```yaml
version: v10.0.0-dev
appVersion: v10.0.0-dev
# ....
dependencies:
  - name: teleport-operator
    version: v10.0.0-dev
    condition: installCRDs
```

It was not really safe to sed without taking care of the indentation, it would have caught any other dependency. Hence I added an anchor and modifierd the `version.mk` logic to edit the anchor instead of the values themselves. While this si not needed for `teleport-kube-agent` I think it's in our best interest to set version on all charts using the same mechanism.

hugoShaka commented on examples/chart/teleport-cluster/charts/teleport-operator/Chart.yaml:
```diff
@@ -0,0 +1,10 @@
+.version: &version "11.0.0-dev"
+
+name: teleport-operator
```
> the operator chart by itself does nothing and won't work if a user just hem installs

I think the subchart name is wrong but the pattern might still be relevant. Most projects are bundling the CRDs directly in the charts as you're suggesting. I think this might be a common antipattern.

Users adopting the operator pattern are deploying multiple operators, with cross-operator dependencies that are even sometimes cyclic. For example:
- The teleport-cluster chart deploys a `ServiceMonitor` CR, managed by the prometheus-operator to be monitored
- The prometheus-operator deploys a `TeleportApplcation` CR to register its Prometheus UIs in Teleport if the `TeleportApplication` CRD exists, else it does nothing.

In this example the operator deployment order matters and we can end up with different final outcomes. This is in contradiction with Kubernete's declarative approach and can lead to degraded features until all operators are restarted at least once.

Such situation can be solved by deploying first all CRDs, then all operators. Such pattern heavily simplifies new cluster deployment. Elastic is doing so with their elastic-cloud operator.

Maybe my past experience with operators is too and most of our users don't care, I'm totally ok with moving the CRDs back to the `teleport-cluster` chart. What's your opinion ?

strideynet commented on operator/sidecar/tbot.go:
```diff
@@ -0,0 +1,254 @@
+}
+
+func (b *Bot) GetClient(ctx context.Context) (auth.ClientI, error) {
+	if !b.running {
```
Are you sure there's no risk of concurrent access to `b.running` ? Shouldn't this have a mutex on it ?

strideynet commented on operator/sidecar/tbot.go:
```diff
@@ -0,0 +1,254 @@
+	// Hack to be able to reuse LoadIdentity functions from tbot
+	// LoadIdentity expects to have all the artifacts required for a bot
+	// We loop over missing artifacts and are loading them from the bot storage to the destination
+	for _, artifact := range identity.GetArtifacts() {
```
Definitely a bit of a hack. Do you know which bits of information that the destination is missing that prevents it from used to load an identity ? I just partially worry that there's a risk that in future something that belongs to the bot is leaked across to the destination, and this leads to an identity being produced that corresponds to the bot itself rather than the impersonated role.

I'd be curious what we could do (as MachineID) to make your life easier here. One thing that definitely comes to mind, is some kind of channel based events subscription, so that consumers of `tbot` can receive a new client/identity immediately when the tbot refreshes (as part of the cycle, or as part of a CA rotation).

hugoShaka replied:
>  Do you know which bits of information that the destination is missing that prevents it from used to load an identity ?

I'm still discovering how we create clients and join clusters but from what I understand:
- The destination is missing both TLS and SSH CA certs, which are required to build a working auth client
- Something called by `LoadIdentity` checks the `tokenhash` artifact, as far as I know this is not useful to get a client for the impersonated role

> I'd be curious what we could do (as MachineID) to make your life easier here. One thing that definitely comes to mind, is some kind of channel based events subscription, so that consumers of tbot can receive a new client/identity immediately when the tbot refreshes (as part of the cycle, or as part of a CA rotation).

In my opinion the following changes would make tbot more embeddable:
- as you suggested, a channel to get notified on renewal would be great, here it would allow me to stop creating new clients for every call which is a waste of resources
- a way to init in-memory destinations more easily and safely (it bit me more often that I want to admit). I think generally users should not have to touch the artifacts at all.
- a method to get an `Identity` for a specific bot destination, this would avoid this dirty hack. In its current state `LoadIdentity` needs all bot-only artifacts `sshcacert`, `tlscacert` and `tokenhash`
- a method on `Identity` yielding a client, `auth.ClientI` or maybe `client.Client`

strideynet replied:
Awesome, that seems to match my notes I prepared this morning on what we could do: https://github.com/gravitational/teleport/issues/14507

I'll link your comment here on there as well, thank you.

hugoShaka commented on operator/controllers/resources/role_controller.go:
```diff
@@ -69,19 +71,38 @@ func (r *RoleReconciler) Delete(ctx context.Context, obj kclient.Object) error {
+	}
+	metadata.Labels[types.OriginLabel] = types.OriginKubernetes
+	(*resource).SetMetadata(metadata)
+}
```
Fixed

hugoShaka replied:
Fixed

hugoShaka commented on operator/apis/resources/v2/user_types.go:
```diff
@@ -0,0 +1,100 @@
+		Version: types.V2,
+		Metadata: types.Metadata{
+			Name:        u.Name,
+			Labels:      u.Labels,
```
I'm not sure what's the impact of doing it here in the `types.go` instead of in the controller.

hugoShaka commented on operator/controllers/resources/user_controller.go:
```diff
@@ -69,23 +71,43 @@ func (r *UserReconciler) Delete(ctx context.Context, obj kclient.Object) error {
+	}
+	metadata.Labels[types.OriginLabel] = types.OriginKubernetes
+	(*resource).SetMetadata(metadata)
 }
```
Fixed

hugoShaka commented on operator/README.md:
```diff
@@ -0,0 +1,158 @@
+
+### Running the operator
+
+Set the `TELEPORT_PROJECT` to the full path to your teleport's project checked out at `marco/plugins-teleport-operator-charts`.
```
Fixed

hugoShaka commented on examples/chart/teleport-cluster/templates/deployment.yaml:
```diff
@@ -205,6 +205,33 @@ spec:
 {{- end }}
+{{ if .Values.operator.enabled }}
+      - name: "operator"
+        image: "{{ .Values.operatorImage }}:{{ if .Values.teleportVersionOverride }}{{ .Values.teleportVersionOverride }}{{ else }}{{ .Chart.Version }}{{ end }}"
```
Done

hugoShaka commented on operator/controllers/resources/reconciler.go:
```diff
@@ -0,0 +1,124 @@
+	}
+
+	if !hasDeletionFinalizer {
+		log.Info("adding finalizer")
```
Note: I reverted the commit as the controller-runtime is already doing it in the default log builder.

marcoandredinis commented on operator/config/crd/bases/resources.teleport.dev_users.yaml:
```diff
@@ -0,0 +1,187 @@
+apiVersion: apiextensions.k8s.io/v1
```
I think we could remove this file (the same goes for the role's one)
CRDs are now stored in the chart itself

marcoandredinis commented on operator/sidecar/sidecar.go:
```diff
@@ -0,0 +1,139 @@
+func (opts *Options) CheckAndSetDefaults() error {
+	if opts.Addr == "" {
+		opts.Addr = DefaultLocalAddr
+
```
Extra empty line :nail_care:
